src/dfx-core/src/config/model/canister_id_store.rs

```diff
@@ -176,19 +295,71 @@ impl CanisterIdStore {
                 canister_name: canister_name.to_string(),
                 canister_id: canister_id.to_string(),
                 cause: e,
-            })
+            })?;
+        if let Some(timestamp) = timestamp {
+            match self.timestamps.get_mut(canister_name) {
+                Some(network_name_to_timestamp) => {
+                    network_name_to_timestamp.insert(network_name.to_string(), timestamp);
+                }
+                None => {
+                    let mut network_name_to_timestamp = NetworkNametoCanisterTimestamp::default();
+                    network_name_to_timestamp.insert(network_name.to_string(), timestamp);
+                    self.timestamps
+                        .insert(canister_name.to_string(), network_name_to_timestamp);
+                }
+            }
+            self.save_timestamps()?;
+        }
+        Ok(())
     }
 
     pub fn remove(&mut self, canister_name: &str) -> Result<(), CanisterIdStoreError> {
         let network_name = &self.network_descriptor.name;
         if let Some(network_name_to_canister_id) = self.ids.get_mut(canister_name) {
             network_name_to_canister_id.remove(network_name);
+            self.save_ids()
+                .map_err(|e| CanisterIdStoreError::RemoveCanisterId {
+                    canister_name: canister_name.to_string(),
+                    cause: e,
+                })?
+        };
+        if let Some(network_name_to_timestamp) = self.timestamps.get_mut(canister_name) {
+            network_name_to_timestamp.remove(network_name);
+            self.save_timestamps()?;
         }
-        self.save_ids()
-            .map_err(|e| CanisterIdStoreError::RemoveCanisterId {
-                canister_name: canister_name.to_string(),
-                cause: e,
-            })
+        Ok(())
+    }
+
+    fn prune_expired_canisters(
+        &mut self,
+        log: &Logger,
+        timeout: &Duration,
+    ) -> Result<(), CanisterIdStoreError> {
+        let network_name = &self.network_descriptor.name;
+        let now = SystemTime::now();
+        let prune_cutoff = now.sub(*timeout);
+
+        let mut canisters_to_prune: Vec<String> = Vec::new();
+        for (canister_name, timestamp) in
+            self.timestamps
+                .iter()
+                .filter_map(|(canister_name, network_to_timestamp)| {
+                    network_to_timestamp
+                        .get(network_name)
+                        .map(|timestamp| (canister_name, timestamp))
+                })
+        {
+            if *timestamp < prune_cutoff {
```

Comment:
Suppose the timeout were 2 seconds.  We'd want to keep canisters created this second (== now-0), the previous second (== now-1), and prune canisters created earlier (<= now-2).
```suggestion
            if *timestamp <= prune_cutoff {
```

It might even be good to add a few seconds to avoid errors related to differences in assumed state right at the 10-minute mark:
- Pruning a canister a few seconds early means creating an additional canister for a few seconds
- Pruning a canister too late could mean trying to deploy to a canister when no longer a controller

Reply:
I changed it to `<=`, but it will be imprecise anyways. We're running update calls, so the time of execution will be at a different time than when we make the call, and the time can be somewhat different between the local machine and the IC time. I think we'll just have to accept that people will run into the corner case occasionally